app/shared/design-system/components/primitives/auto-input.tsx

```diff
@@ -0,0 +1,217 @@
+							const {selectionStart, selectionEnd, selectionDirection} = currentTarget;
+							currentTarget.value = valueTokens.map(({tokenValue}) => tokenValue).join('');
+							currentTarget.setSelectionRange(selectionStart, selectionEnd, selectionDirection);
+							if (selectTarget.dataset.selected) selectTarget.blur();
+						};
+						break;
+					}
+				}
+				break;
+			}
+			case 'pointerup': {
+				this.#ownerDocument.documentElement.style.cursor = '';
+				this.#activeTarget = null;
+				currentTarget.state = null;
+				break;
+			}
+			case 'pointermove': {
+				const measureMetrics = currentTarget.state || this.#measureEvent(currentTarget.value, event, currentTarget);
+				currentTarget.props = measureMetrics;
+				currentTarget.style.cursor = '';
+				const {valueOffset, valueTokens} = measureMetrics;
+				if (~valueOffset) {
+					const activeToken = valueTokens[valueOffset];
+					switch (activeToken.tokenType) {
+						case 'number': {
+							if (this.#activeTarget) {
+								event.preventDefault();
+								const {movementY} = event;
+								let tokenValue = parseFloat(activeToken.tokenValue) - movementY;
+								if (tokenValue % 1 !== 0 || movementY % 1 !== 0) tokenValue = tokenValue.toPrecision(Math.abs(tokenValue).toString().indexOf('.') + 2);
+								activeToken.tokenValue = tokenValue;
+								let {selectionStart, selectionEnd} = currentTarget;
+								if (selectionStart > selectionEnd) [selectionStart, selectionEnd] = [selectionEnd, selectionStart];
+								currentTarget.value = valueTokens.map(({tokenValue}) => tokenValue).join('');
+								while (/[+-.\d]/.test(currentTarget.value[selectionStart - 1])) selectionStart--;
+								selectionEnd = selectionStart;
+								while (/[.\d]/.test(currentTarget.value[++selectionEnd]));
+								currentTarget.setSelectionRange(selectionStart, selectionEnd);
+							}
+							currentTarget.style.cursor = this.#resizeCursor;
+							break;
+						}
+						case 'unit': {
+							currentTarget.style.cursor = this.#pointerCursor;
+							break;
+						}
+						default: {
+							currentTarget.style.cursor = '';
+						}
+					}
+				}
+				break;
+			}
+			case 'keydown': {
+				switch (event.code) {
+					case 'ArrowUp': case 'ArrowDown': {
+						let {selectionStart, selectionEnd} = currentTarget;
+						if (selectionStart > selectionEnd) [selectionStart, selectionEnd] = [selectionEnd, selectionStart];
+						const {pixelPadding} = currentTarget.props;
+						const currentTargetValue = currentTarget.value;
+						const currentSelectionValue = currentTargetValue.substring(currentTarget.selectionStart, currentTarget.selectionEnd);
+						let offsetX = this.#canvasContext.measureText(currentTargetValue.substring(0, currentTarget.selectionStart)).width;
+						if (pixelPadding) offsetX += pixelPadding;
+						if (!isNaN(parseFloat(currentSelectionValue))) offsetX += 0.25;
```

Comment:
Very few people know: `!isNaN` is very hard to read, because of `!i`, instead a much better thing to do for readability is `if (isNaN() === false)`